#!/usr/bin/env python

################################################################
#
# summarizeCGPMresults.py
#
# Description:  This program parses a data file produced by queryCGPMresults.py,
#   which contains a slice through output generated by compareGeneProfiles_main.py
#   and postProcessCGPMresults.py, run in high-throughput using wrapper scripts.
#   Data were queried from a list of directories using queryCGPMresults.py.
#   summarizeCGPMresults.py extracts the genecall information for a designated reference
#   and lists it (one gene call per line) along with the corresponding (matching) gene
#   call information for the other strains/species against which PAK1 was compared.
#
# Programmer's Notes
#   This code was modified from a previously rather hard-coded version first written
#   in a big hurry due to extreme time constraints. This version should work properly
#   for an abritrary input reference genome and any number of other genomes to which
#   the reference was compared.
#
# Programmer: CEZhou
#    18 Nov 2013: begin
#    11 March 2014: Removed hard-coding and tested against PAK1 NxN data set
#    12 March 2014: Output goes to .out file rather than standard out
#*** 28 March 2014: Code has been added to accommodate reporting of loners
#          Note: this code (and query codes) should be re-cast using classes; right now it's a bit of
#          a jumble.  Also, using classes it should be simpler/easier to pull out
#          specific subsets of data, sorted.
#
################################################################
'''
Copyright (c) 2015, Lawrence Livermore National Security, LLC. Produced at the Lawrence Livermore National Laboratory. Written by Carol L. Ecale Zhou, zhou4@llnl.gov; carol.zhou@comcast.net. CODE-OCEC-15-045 All rights reserved. This file is part of Qspp. Please also read this link – Our Notice and GNU General Public License. 
This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License (as published by the Free Software Foundation) version 2, dated June 1991. 
This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and conditions of the GNU General Public License for more details. 
You should have received a copy of the GNU General Public License along with this program; if not, write to the Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA 
OUR NOTICE AND TERMS AND CONDITIONS OF THE GNU GENERAL PUBLIC LICENSE 
Our Preamble Notice 
A. This notice is required to be provided under our contract with the U.S. Department of Energy (DOE). This work was produced at the Lawrence Livermore National Laboratory under Contract No. DE-AC52-07NA27344 with the DOE. 
B. Neither the United States Government nor Lawrence Livermore National Security, LLC nor any of their employees, makes any warranty, express or implied, or assumes any liability or responsibility for the accuracy, completeness, or usefulness of any information, apparatus, product, or process disclosed, or represents that its use would not infringe privately-owned rights. 
C. Also, reference herein to any specific commercial products, process, or services by trade name, trademark, manufacturer or otherwise does not necessarily constitute or imply its endorsement, recommendation, or favoring by the United States Government or Lawrence Livermore National Security, LLC. The views and opinions of authors expressed herein do not necessarily state or reflect those of the United States Government or Lawrence Livermore National Security, LLC, and shall not be used for advertising or product endorsement purposes. 
The precise terms and conditions for copying, distribution and modification follows. 
GNU Terms and Conditions for Copying, Distribution, and Modification 
_ 0. This License applies to any program or other work which contains a notice placed by the copyright holder saying it may be distributed under the terms of this General Public License. The “Program,” below, refers to any such program or work, and a “work based on the Program” means either the Program or any derivative work under copyright law: that is to say, a work containing the Program or a portion of it, either verbatim or with modifications and/or translated into another language. (Hereinafter, translation is included without limitation in the term “modification”.) Each licensee is addressed as “you.” 
_ Activities other than copying, distribution and modification are not covered by this License; they are outside its scope. The act of running the Program is not restricted, and the output from the Program is covered only if its contents constitute a work based on the Program (independent of having been made by running the Program). Whether that is true depends on what the Program does. 
_ 1. You may copy and distribute verbatim copies of the Program’s source code as you receive it, in any medium, provided that you conspicuously and appropriately publish on each copy an appropriate copyright notice and disclaimer of warranty; keep intact all the notices that refer to this License and to the absence of any warranty; and give any other recipients of the Program a copy of this License along with the Program. 
_ You may charge a fee for the physical act of transferring a copy, and you may at your option offer warranty protection in exchange for a fee. 
2 You may modify your copy or copies of the Program or any portion of it, thus forming a work based on the Program, and copy and distribute such modifications or work under the terms of Section 1 above, provided that you also meet all of these conditions: 
a) You must cause the modified files to carry prominent notices stating that you changed the files and the date of any change. 
b) You must cause any work that you distribute or publish, that in whole or in part contains or is derived from the Program or any part thereof, to be licensed as a whole at no charge to all third parties under the terms of this License. 
c) If the modified program normally reads commands interactively when run, you must cause it, when started running for such interactive use in the most ordinary way, to print or display an announcement including an appropriate copyright notice and a notice that there is no warranty (or else, saying that you provide a warranty) and that users may redistribute the program under these conditions, and telling the user how to view a copy of this License. (Exception: if the Program itself is interactive but does not normally print such an announcement, your work based on the Program is not required to print an announcement.) 
These requirements apply to the modified work as a whole. If identifiable sections of that work are not derived from the Program, and can be reasonably considered independent and separate works in themselves, then this License, and its terms, do not apply to those sections when you distribute them as separate work. But when you distribute the same section as part of a whole which is a work based on the Program, the distribution of the whole must be on the terms of this License, whose permissions for other licensees extend to the entire whole, and thus to each and every part regardless of who wrote it. 
Thus, it is not the intent of this section to claim rights or contest your rights to work written entirely by you; rather, the intent is to exercise the right to control the distribution of derivative or collective works based on the Program. 
In addition, mere aggregation of another work not based on the Program with the Program (or with a work based on the Program) on a volume of a storage or distribution medium does not bring the other work under the scope of this License. 
3. You may copy and distribute the Program (or a work based on it, under Section 2) in object code or executable form under the terms of Sections 1 and 2 above provided that you also do one of the following: 
a) Accompany it with the complete corresponding machine-readable source code, which must be distributed under the terms of Sections 1 and 2 above on a medium customarily used for software interchange; or, 
b) Accompany it with a written offer, valid for at least three years, to give any third party, for a charge no more than your cost of physically performing source distribution, a complete machine-readable copy of the corresponding source code, to be distributed under the terms of Sections 1 and 2 above on a medium customarily used for software interchange; or, 
c) Accompany it with the information you received as to the offer to distribute corresponding source code. (This alternative is allowed only for noncommercial distribution and only if you received the program in object code or executable form with such an offer, in accord with Subsection b above.) 
The source code for a work means the preferred form the work for making modifications to it. For an executable work, complete source code means all the source code for all modules it contains, plus any associated interface definition files, plus the scripts used to control compilation and installation of the executable. However, as a special exception, the source code distributed need not include anything that is normally distributed (in either source or binary form) with the major components (compiler, kernel, and so on) of the operating system on which the executable runs, unless that component itself accompanies the executable. 
If distribution of executable or object code is made by offering access to copy from a designated place, then offering equivalent access to copy the source code from the same place counts as distribution of the source code, even though third parties are not compelled to copy the source along with the object code. 
4. You may not copy, modify, sublicense, or distribute the Program except as expressly provided under this License. Any attempt otherwise to copy, modify, sublicense or distribute the 
Program is void, and will automatically terminate your rights under this License. However, parties who have received copies, or rights, from you under this License will not have their licenses terminated so long as such parties remain in full compliance. 
2 You are not required to accept this License, since you have not signed it. However, nothing else grants you permission to modify or distribute the Program or its derivative works. These actions are prohibited by law if you do not accept this License. Therefore, by modifying or distributing the Program (or any work based on the Program), you indicate your acceptance of this License to do so, and all its terms and conditions for copying, distributing or modifying the Program or works based on it. 
3 Each time you redistribute the Program (or any work based on the Program), the recipient automatically receives a license from the original licensor to copy, distribute or modify the Program subject to these terms and conditions. You may not impose any further restrictions on the recipients’ exercise of the rights granted herein. You are not responsible for enforcing compliance by third parties to this License. 
4 If, as a consequence of a court judgment or allegation of patent infringement or for any other reason (not limited to patent issues), conditions are imposed on you (whether by court order, agreement or otherwise) that contradict the conditions of this License, they do not excuse you from the conditions of this License. If you cannot distribute so as to satisfy simultaneously your obligations under this License and any other pertinent obligations, then as a consequence you may not distribute the Program at all. For example, if a patent license would not permit royalty-free redistribution of the Program by all those who receive copies directly or indirectly through you, then the only way you could satisfy both it and this License would be to refrain entirely from distribution of the Program. 
If any portion of this section is held invalid or unenforceable under any particular circumstance, the balance of the section is intended to apply and the section as a whole is intended to apply in other circumstances. 
It is not the purpose to this section to induce you to infringe any patents or other property right claims or to contest validity of any such claims; this section has the sole purpose of protecting the integrity of the free software distribution system, which is implemented by public license practices. Many people have made generous contributions to the wide range of software distributed through that system in reliance on consistent application of that system; it is up to the author/donor to decide if he or she is willing to distribute software through any other system and a licensee cannot impose that choice. 
This section is intended to make thoroughly clear what is believed to be a consequence of the rest of this License. 
1 If the distribution and/or use of the Program is restricted in certain countries either by patents or by copyrighted interfaces, the original copyright holder who places the Program under this License may add an explicit geographical distribution limitation excluding those countries, so that distribution is permitted only in or among countries not thus excluded. In such case, this License incorporates the limitation as if written in the body of this License. 
9. The Free Software Foundation may publish revised and/or new versions of the General Public License from time to time. Such new versions will be similar in spirit to the present version, but may differ in detail to address new problems or concerns. 
Each version is given a distinguishing version number. If the Program specifies a version number of this License which applies to it and “any later version,” you have the option of following the terms and conditions either of that version of any later version published by the Free Software Foundation. If the Program does not specify a version number of this License, you may choose any version ever published by the Free Software Foundation. 
2 If you wish to incorporate parts of the Program into other free programs whose distribution conditions are different, write to the author to ask for permission. For software which is copyrighted by the Free Software Foundation, write to the Free Software Foundation; we sometimes make exceptions for this. Our decision to grant permission will be guided by the two goals of preserving the free status of all derivatives of our free software and or promoting the sharing and reuse of software generally. 
NO WARRANTY 
11. BECAUSE THE PROGRAM IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY FOR THE PROGRAM, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES PROVIDE THE PROGRAM “AS IS” WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND 
FITNESS FOR A PARTICULAR PURPOSE. THE ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF 
THE PROGRAM IS WITH YOU. SHOULD THE PROGRAM PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL NECESSARY SERVICING, REPAIR OR CORRECTION. 
2 IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR REDISTRIBUTE THE PROGRAM AS PERMITTED ABOVE, BE LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL INCIDENTAL OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE THE PROGRAM (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A FAILURE OF THE PROGRAM TO OPERATE WITH ANY OTHER PROGRAMS), EVEN IF SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGES. 
END OF TERMS AND CONDITIONS
'''

import sys
import re
import copy
from time import strftime
from time import gmtime
import os

dateTime = strftime("%Y:%m:%d::%H:%M:%S", gmtime())  # get time down to seconds
CODE_BASE_DIR = os.environ["CODE_BASE_DIR"]
##### FILES

infile = ""  # user provided
outfile = os.path.join(CODE_BASE_DIR, "summarizeCGPMresults.out")  # default
logfile = os.path.join(CODE_BASE_DIR, "summarizeCGPMresults.log")
LOGFILE = open(logfile,"w")
LOGFILE.write("%s%s\n" % ("Begin log file ",dateTime))
##### PATTERNS

#p_pattern = re.compile(' ')

##### DECLARATIONS

fileError = False

##### CONSTANTS

#LONER = False  # if query was for loner genes, no corresponding gene on other genome,
LONER = True   # which affects some code segments

PATH_FILE = int(5)      # identifies which word from input string from which to extract path/file
#ORGANISM_DIR = int(6)  # identifies which segment of the path/file from which to extract strains' dir
ORGANISM_DIR = ""       # change to penultimate /-delimited field of path/filename (compute below)
# The following are indexes to the fields in tabbed data line from input file containing mutual and singular hits
POSITION    = int(0)    # position at which hit was placed along reference genome (ie, sort position)
GENOME_TYPE = int(1)    # genome1|2 followed by "_" followed by  mutual|singular|loner
G1HEADER    = int(2)    # genome1 header
G1CONTIG    = int(3)
G1ANNOT     = int(4)    # list of annotations (as string; not being handled here)
G2HEADER    = int(5)    # genome2 header
G2CONTIG    = int(6)
G2ANNOT     = int(7)    # list of annotations (as string; not being handled here)
G1LENGTH    = int(8)    # length of G1 gene call
G2LENGTH    = int(9)    # length of G2 gene call
INTERPRETATION = int(10) # listing of interpretations (originally added by postProcessCGMP.py)
# The following indexes apply when the queryCGPMresults file contains loners
HEADER      = int(2)
CONTIG      = int(3)
ANNOTATION  = int(4)
LENGTH      = int(5)

HELP_STRING = "This script takes as input the name of a reference genome (ie, \'Sterne\') plus a file generated by queryCGPMresults.py, and produces output comprising genetic feature records for each genetic feature selected from the query. For each genetic feature of interest, the corresponding feature is identified for each genome that was compared. Fields displayed are: genome, header, annotation, length, and interpretation.  For more information, type: python summarizeCGPMresults.py usage|input\n"

USAGE_STRING = "Usage: summarizeCGPMresults.py <referenceGenome> <infile> <outfile>\nOutfile is optional. Default output file is summarizeCGPMresults.out"

INPUT_STRING = "Input: name of reference genome (ie, \'Sterne\') plus file containing query results generated by code queryCGPMresults.py. Caution: the reference genome should be spelled exactly as it is in the directory structure containing the comparison data sets.\n"

ACCEPTABLE_ARG_COUNT = (3,4) # 4 if user is providing name of output file 

#DEBUG = True
#DEBUG = False

##### VARIABLES

reference = "error"  # provided by user
otherGenome = "error"

myDict = {
    "myStuff" : "example",
    }

##### FUNCTIONS

def GetGeneData(genomeNumber, fields, geneDataInstance):
    # genomeNumber is '1' or '2'
    # fields is
    # geneDataInstance
    header                             = fields[G1HEADER]
    (locusTag,strand,start,end,extra)  = header.split('/')
    geneDataInstance["header"]         = header
    geneDataInstance["cds"]            = locusTag
    (junk,cdsNumber)                   = locusTag.split('cds')
    geneDataInstance["cdsNumber"]      = cdsNumber
    geneDataInstance["strand"]         = strand
    geneDataInstance["start"]          = start
    geneDataInstance["end"]            = end
    geneDataInstance["annotation"]     = fields[G1ANNOT]
    lengthString                       = fields[G1LENGTH]
    (junk, geneDataInstance["length"]) = lengthString.split('=')
    geneDataInstance["interpretation"] = fields[INTERPRETATION]
    return (geneDataInstance) 

##### GET INPUT PARAMETERS

argCount = len(sys.argv)
if argCount in ACCEPTABLE_ARG_COUNT:
    match = re.search("help", sys.argv[1].lower())
    if match:
        print HELP_STRING
        LOGFILE.close(); exit(0)
    match = re.search("input", sys.argv[1].lower())
    if match:
        print INPUT_STRING
        LOGFILE.close(); exit(0)
    match = re.search("usage", sys.argv[1].lower())
    if match:
        print USAGE_STRING
        LOGFILE.close(); exit(0)

    if argCount == 3:
        reference = sys.argv[1]
        infile = sys.argv[2] 
    elif argCount == 4:
        reference = sys.argv[1]
        infile = sys.argv[2] 
        outfile = sys.argv[3]
    else:
        print USAGE_STRING
        LOGFILE.close(); exit(0)
else:
    print USAGE_STRING
    LOGFILE.write("%s\n" % ("Incorrect number of command-line arguments provided"))
    LOGFILE.close(); exit(0)

# Open files
try:
    INFILE = open(infile,"r")
except IOError as e:
    fileError = True
    print e
try:
    OUTFILE = open(outfile,"w")
except IOError as e:
    fileError = True
    print e
if fileError:
    LOGFILE.close(); exit(0)


##### BEGIN MAIN 

# Additional Variables

genome = ""; hitType = ""
genome1 = ""; genome2 = ""; g1length = ""; g2length = ""
g1header = ""; g2header = ""; g1annot = ""; g2annot = ""
position = ""; genome_hitType = ""; interpretation = ""
words = []; pathFile = ""; segments = []; fields = []; lengthElements = []
junk = ""
cdsNumber = 0

referenceGeneList    = []  # will capture a non-redundant list of genes from input data
referenceHomologList = []  # will hold the lists of compared-genome genetic features data, indexed by order corresponding to referenceGeneList
comparedGenomesList  = []  # holds a non-redundant list of genomes that were compared to reference
geneExclusions       = {}  # key = header, value = exclusionList object 
exclusionList        = []  # holds list of genomes that reference gene is unique with respect to
geneAnnotation       = {}  # holds annotations for genes, key is compoundKey = header // contig
starts               = {}  # holds start locations (for sorting)

geneData = {
    "genome"         : "",  # genome with a corresponding genetic feature or gene 
    "header"         : "",
    "cds"            : "",
    "cdsNumber"      : 0,
    "strand"         : "",
    "start"          : 0,
    "end"            : 0,
    "contig"         : "",
    "annotation"     : "",
    "length"         : "",
    "interpretation" : "",
    }

homologData = {}  # holds a set of geneData objects

##### Read through input data file twice, first to collect all genetic features (ie, genes) for
##### the designated reference genome plus a non-redundant list of the genomes to which the
##### reference was compared, and then again to capture and organize all of the corresponding
##### genetic features from the compared genomes.
    
##### First pass:  collect all 1) compared genomes and 2) gene calls for reference genome into lists

fLines = INFILE.read().splitlines()
for line in fLines:
    match = re.search('Gene Set 1', line)        # Process each binary comparison
    words = []
    if match:                                    # Capture the 1st genome in the pair                      
        words = line.split(' ')
        pathFile = words[PATH_FILE]
        segments = pathFile.split('/')           # Split apart fully qualified path/filename
        genome1 = segments[-2]                   # sub-dir name for this genome's data is in penultimate position
        if genome1 not in comparedGenomesList:   # capture name of 1st compared genome
            if genome1 != reference:             # exclude reference from this list
                comparedGenomesList.append(genome1)
    match = re.search('Gene Set 2', line)        # Process each binary comparison
    if match:                                    # Capture the 2nd genome in the pair
        words = line.split(' ')
        pathFile = words[PATH_FILE]
        segments = pathFile.split('/')           # Split apart fully qualified path/filename
        genome2 = segments[-2]                   # sub-dir name for this genome's data is in penultimate position
        if genome2 not in comparedGenomesList:   # capture name of 2nd compared genome
            if genome2 != reference:             # exclude reference
                comparedGenomesList.append(genome2)

    match = re.search('^\d', line)               # Compile a list of all genes from reference
    if match:                                    # Line contains genetic feature data 
        fields = line.split('\t')
        (genome, hitType) = fields[GENOME_TYPE].split('_')
        if genome1 == reference:                 # Unless genome1 or genome2 is reference, don't bother 
            if not LONER:
                if fields[G1HEADER] not in referenceGeneList: # capture genetic feature / gene
                    referenceGeneList.append(fields[G1HEADER])
            if LONER:  # add gene to list only if this is a genome1_loner record
                if genome == 'genome1':
                    if fields[HEADER] not in referenceGeneList:  
                        referenceGeneList.append(fields[HEADER])   #*** ??? compoundKey here?
        if genome2 == reference:                 # Unless genome1 or genome2 is reference, don't bother
            if not LONER:
                if fields[G2HEADER] not in referenceGeneList: # capture genetic feature / gene
                    referenceGeneList.append(fields[G2HEADER]) 
            if LONER:  # add gene to list only if this is a genome2_loner record
                if genome == 'genome2':
                    if fields[HEADER] not in referenceGeneList:
                        referenceGeneList.append(fields[HEADER])

INFILE.close()

### The following section is needed only for mutual and singular hits
### For each reference gene found, create a data record and insert into a list in same order 

if not LONER:  # For mutual and singular hit data
    locusTag = ""; strand = ""; start = ""; end = ""; extra = ""
    headerFields = []
    for header in referenceGeneList:
        headerFields = header.split('/')
        locusTag = headerFields[0]
        strand   = headerFields[1]
        start    = headerFields[2]
        end      = headerFields[3]
        (junk,cdsNumber) = locusTag.split('cds')
        nextGeneData = copy.deepcopy(geneData)   # create empty data structure
        nextGeneData["header"]    = header
        nextGeneData["cdsNumber"] = cdsNumber
        nextGeneData["strand"]    = strand
        nextGeneData["start"]     = start
        nextGeneData["end"]       = end
        newHomologData = copy.deepcopy(homologData)
        newHomologData[reference] = nextGeneData # first item in the homolog list is the reference's own gene data
        referenceHomologList.append(newHomologData)  # place empty data structure; same index as header in homologGeneList 

### Second pass:  for each reference gene call, complete geneData record based on infile data

INFILE = open(infile,"r")
fLines = INFILE.read().splitlines()

if LONER:  # For detected unique genes on reference genome

    # First, create an empty exclusion list for each reference gene
    print "Number of reference genes:", len(referenceGeneList)
    print "Number of compared genomes:", len(comparedGenomesList)
    for header in referenceGeneList:
        newExclusionList = copy.deepcopy(exclusionList)
        geneExclusions[header] = newExclusionList

    # Second, determine genome1 and genome2 for next comparison set, then record genome exclusions for each reference gene
    for line in fLines:
        match = re.search('Gene Set 1', line)
        if match:
            words = line.split(' ')
            pathFile = words[PATH_FILE]
            segments = pathFile.split('/')
            genome1 = segments[-2]
        match = re.search('Gene Set 2', line)
        if match:
            words = line.split(' ')
            pathFile = words[PATH_FILE]
            segments = pathFile.split('/')
            genome2 = segments[-2]

        # For each reference loner gene in this list, record that the gene is unique wrt the other genome
        segments = []
        match = re.search('^\d', line)
        if match:
            fields = line.split('\t')
            (genome,hitType) = fields[GENOME_TYPE].split('_')
            if genome == 'genome1' and genome1 == reference:
                header = fields[HEADER]
                contig = fields[CONTIG]
                annot  = fields[ANNOTATION]
                contig_annotation = fields[CONTIG] + '_/_' + fields[ANNOTATION]
                geneExclusions[header].append(genome2) 
                geneAnnotation[header] = contig_annotation 
                segments = header.split('/')
                starts[header] = segments[2]
                #geneExclusions[header].append(genome2) 
            if genome == 'genome2' and genome2 == reference:
                header = fields[HEADER]
                contig = fields[CONTIG]
                annot  = fields[ANNOTATION]
                contig_annotation = fields[CONTIG] + '_/_' + fields[ANNOTATION]
                geneExclusions[header].append(genome1)
                geneAnnotation[header] = contig_annotation 
                segments = header.split('/')
                starts[header] = segments[2]
                #geneExclusions[header].append(genome1)

if not LONER:  # For mutual and singular hit data, record for each reference gene the corresponding genes from other genomes
    for line in fLines:
    
        # Determine which genome was #1 vs #2 
        match = re.search('Gene Set 1', line)
        if match:
            words = line.split(' ')
            pathFile = words[PATH_FILE]
            segments = pathFile.split('/')   # split apart fully qualified path/filename
            genome1 = segments[-2]           # name of organism's subdir is in penultimate position
        match = re.search('Gene Set 2', line)
        if match:
            words = line.split(' ')
            pathFile = words[PATH_FILE]
            segments = pathFile.split('/')   # split apart fully qualified path/filename
            genome2 = segments[-2]           # name of organism's subdir is in penultimate position

        # Gather a list of the genes that were selected from the genome of interest
        match = re.search('^\d', line)       # genetic-feature data lines begin with the position number
        if match:  # First, get handle on array index where geneData is kept; then store data
            referenceGeneData  = copy.deepcopy(geneData)
            otherGeneData      = copy.deepcopy(geneData)
            fields = line.split('\t')
            (genome,hitType) = fields[GENOME_TYPE].split('_')   

            if genome1 == reference:  # Unless genome1 or genome2 is reference, don't bother
                otherGenome = genome2 
                #*** Note: indexing by header is not necessarily unique (ie could be cds, start, end all same on different contigs, though highly unlikely)
                insertionPoint = referenceGeneList.index(fields[G1HEADER])  # get array index; corresponding locations!

                #*** Commented out code replaces redundant code below, but needs testing first !!!
                # Get Genome 1 data, put to reference data frame
                # referenceGeneData = GetGeneData(1,fields,referenceGeneData)
                # referenceHomologList[insertionPoint][reference] = referenceGeneData
                header                              = fields[G1HEADER]
                (locusTag,strand,start,end,extra)   = header.split('/') 
                referenceGeneData["header"]         = header
                referenceGeneData["cds"]            = locusTag
                (junk,cdsNumber)                    = locusTag.split('cds')
                referenceGeneData["cdsNumber"]      = cdsNumber
                referenceGeneData["strand"]         = strand
                referenceGeneData["start"]          = start
                referenceGeneData["end"]            = end
                referenceGeneData["annotation"]     = fields[G1ANNOT]
                lengthString                        = fields[G1LENGTH]
                (junk, referenceGeneData["length"]) = lengthString.split('=')
                referenceGeneData["interpretation"] = fields[INTERPRETATION]
                referenceHomologList[insertionPoint][reference] = referenceGeneData

                # Get Genome2 data, put to other genome's data frame
                # otherGeneData = GetGeneData(2,fields,referenceGeneData)
                # referenceHomologList[insertionPoint][otherGenome] = referenceGeneData
                header                            = fields[G2HEADER]
                (locusTag,strand,start,end,extra) = header.split('/') 
                otherGeneData["header"]           = header
                otherGeneData["cds"]              = locusTag
                (junk,cdsNumber)                  = locusTag.split('cds')
                otherGeneData["cdsNumber"]        = cdsNumber
                otherGeneData["strand"]           = strand
                otherGeneData["start"]            = start
                otherGeneData["end"]              = end
                otherGeneData["annotation"]       = fields[G2ANNOT]
                lengthString                      = fields[G2LENGTH]
                (junk, otherGeneData["length"])   = lengthString.split('=')
                otherGeneData["interpretation"]   = fields[INTERPRETATION]
                referenceHomologList[insertionPoint][otherGenome] = otherGeneData

            elif genome2 == reference:  # Unless genome1 or genome2 is reference, don't bother
                otherGenome = genome1
                insertionPoint = referenceGeneList.index(fields[G2HEADER])  # get array index; corresponding locations!

                # Get Genome 2 data, put to reference data frame
                # referenceGeneData = GetGeneData(2,fields,referenceGeneData)
                # referenceHomologList[insertionPoint][reference] = referenceGeneData
                header                              = fields[G2HEADER]
                (locusTag,strand,start,end,extra)   = header.split('/') 
                referenceGeneData["header"]         = header
                referenceGeneData["cds"]            = locusTag
                (junk,cdsNumber)                    = locusTag.split('cds')
                referenceGeneData["cdsNumber"]      = cdsNumber
                referenceGeneData["strand"]         = strand
                referenceGeneData["start"]          = start
                referenceGeneData["end"]            = end
                referenceGeneData["annotation"]     = fields[G2ANNOT]
                lengthString                        = fields[G2LENGTH]
                (junk, referenceGeneData["length"]) = lengthString.split('=')
                referenceGeneData["interpretation"] = fields[INTERPRETATION]
                referenceHomologList[insertionPoint][reference] = referenceGeneData

                # Get Genome1 data, put to other genome's data frame
                # otherGeneData = GetGeneData(1,fields,referenceGeneData)
                # referenceHomologList[insertionPoint][otherGenome] = referenceGeneData
                header                            = fields[G1HEADER]
                (locusTag,strand,start,end,extra) = header.split('/') 
                otherGeneData["header"]           = header
                otherGeneData["cds"]              = locusTag
                (junk,cdsNumber)                  = locusTag.split('cds')
                otherGeneData["cdsNumber"]        = cdsNumber
                otherGeneData["strand"]           = strand
                otherGeneData["start"]            = start
                otherGeneData["end"]              = end
                otherGeneData["annotation"]       = fields[G1ANNOT]
                lengthString                      = fields[G1LENGTH]
                (junk, otherGeneData["length"])   = lengthString.split('=')
                otherGeneData["interpretation"]   = fields[INTERPRETATION]
                referenceHomologList[insertionPoint][otherGenome] = otherGeneData

    # Check results

    listSize = len(referenceGeneList)
    OUTFILE.write("%s%s\n" % ("Number of entries in referenceGeneList",len(referenceGeneList)))
    OUTFILE.write("%s\n" % ("Genome\tHeader\tAnnotation\tLength\tInterpretation"))
    #print "Number of entries in referenceGeneList:", len(referenceGeneList)
    #print "Genome\tHeader\tAnnotation\tLength\tInterpretation\t"
    for i in xrange(0,listSize-1):
        OUTFILE.write("%s%s\n" % ("Gene No. ", i+1))
        #print "Gene No.", i+1
        if "header" in referenceHomologList[i][reference]:   #
            header = referenceHomologList[i][reference]["header"]
            annotation = referenceHomologList[i][reference]["annotation"]
            length = referenceHomologList[i][reference]["length"]
            interpretation = referenceHomologList[i][reference]["interpretation"]
            OUTFILE.write("%s\t%s\t%s\t%s\t%s\n" % (reference,header,annotation,length,interpretation))
            #print reference, "\t", header, '\t', annotation, '\t', length, '\t', interpretation 
            if "header" in referenceHomologList[i][reference]:
                for otherGenome in comparedGenomesList:
                    if otherGenome in referenceHomologList[i]:
                        header = referenceHomologList[i][otherGenome]["header"]
                        annotation = referenceHomologList[i][otherGenome]["annotation"]
                        length = referenceHomologList[i][otherGenome]["length"]
                        interpretation = referenceHomologList[i][otherGenome]["interpretation"]
                        OUTFILE.write("%s\t%s\t%s\t%s\t%s\n" % (otherGenome,header,annotation,length,interpretation))
                        #print otherGenome, "\t", header, '\t', annotation, '\t', length, '\t', interpretation 
                    else:
                        OUTFILE.write("%s%s\n" % (otherGenome,": no record found"))
                        #print otherGenome, ": no record found"
        else:
            OUTFILE.write("%s%s\n" % (reference,": no record found"))
            #print reference, ": no record found"

        OUTFILE.write("%s\n" % ("################################################"))
        #print "###################################################"

if LONER:  # Print out a list of reference genome loners and the compared genomes that they are unique with respect to
    # First, alphabetize the compared genomes list and write header line in output file
    comparedGenomesList.sort()
    OUTFILE.write("%s\t%s\t%s\t%s" % ("Reference genetic feature","Start","Contig","Annotation"))
    for genome in comparedGenomesList:
        OUTFILE.write("\t%s" % (genome))
    OUTFILE.write("\n")

    # Next, prepare to write chart of reference genes vs. compared genomes
    for header in geneExclusions:
        printArray = []
        for j in xrange(0,len(comparedGenomesList)):
             printArray.append('')
        (contig,annotation) = geneAnnotation[header].split('_/_')
        OUTFILE.write("%s\t%s\t%s\t%s" % (header,starts[header],contig,annotation))
        for genome in geneExclusions[header]:
            i = comparedGenomesList.index(genome)
            printArray[i] = 'X'
        #OUTFILE.write("%s" % printArray)
        for item in printArray:
            OUTFILE.write("\t%s" % (item))  
        OUTFILE.write("\n")
            
##### CLEAN UP

INFILE.close()
OUTFILE.close()
dateTime = "0:0:0:0:0:0"
dateTime = strftime("%Y:%m:%d::%H:%M:%S", gmtime())  # get time down to seconds
#dateTime = str(datetime.datetime.now().time()) # get time down to sub-seconds
#dateTime = os.popen('date')
LOGFILE.write("%s%s\n" % ("Processing complete ",dateTime))
LOGFILE.close()
